src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/OOPTagHelperResolver.cs

```diff
@@ -105,14 +109,58 @@ protected virtual async Task<TagHelperResolutionResult> ResolveTagHelpersOutOfPr
                 return null;
             }
 
+            if (!_resultCache.TryGetId(projectSnapshot.FilePath, out var lastResultId))
+            {
+                lastResultId = -1;
+            }
+
             var projectHandle = new ProjectSnapshotHandle(projectSnapshot.FilePath, projectSnapshot.Configuration, projectSnapshot.RootNamespace);
-            var result = await remoteClient.TryInvokeAsync<IRemoteTagHelperProviderService, TagHelperResolutionResult>(
+            var result = await remoteClient.TryInvokeAsync<IRemoteTagHelperProviderService, TagHelperDeltaResult>(
                 workspaceProject.Solution,
-                (service, solutionInfo, innerCancellationToken) => service.GetTagHelpersAsync(solutionInfo, projectHandle, factory?.GetType().AssemblyQualifiedName, innerCancellationToken),
+                (service, solutionInfo, innerCancellationToken) => service.GetTagHelpersDeltaAsync(solutionInfo, projectHandle, factory?.GetType().AssemblyQualifiedName, lastResultId, innerCancellationToken),
                 cancellationToken
             );
 
-            return result.HasValue ? result.Value : null;
+            if (!result.HasValue)
+            {
+                return null;
+            }
+
+            var tagHelpers = ConsumeTagHelpersFromDelta(projectSnapshot.FilePath, lastResultId, result.Value);
+
+            var resolutionResult = new TagHelperResolutionResult(tagHelpers, diagnostics: Array.Empty<RazorDiagnostic>());
+            return resolutionResult;
+        }
+
+        // Protected virtual for testing
+        protected virtual IReadOnlyList<TagHelperDescriptor> ConsumeTagHelpersFromDelta(string projectFilePath, int lastResultId, TagHelperDeltaResult deltaResult)
+        {
+            if (!_resultCache.TryGet(projectFilePath, lastResultId, out var tagHelpers))
+            {
+                // We most likely haven't made a request to the server yet so there's no "delta" to apply
+                tagHelpers = Array.Empty<TagHelperDescriptor>();
+
+                if (deltaResult.Delta)
+                {
+                    // We somehow failed to retrieve a cached object yet the server was able to apply a delta. This
+                    // is entirely unexpected and means the server & client are catastrophically de-synchronized.
+                    throw new InvalidOperationException("This should never happen. Razor server & client are de-synchronized. Tearing down");
```

Comment:
You're not wrong; however, the alternate is that if we can't find a comparable "Result" to apply the delta too we'll effectively be generating partial TagHelpers for customers because our delta will be a "delta" result without anything to apply said delta to. In practice the caching bit can't happen today only because we only request new TagHelpers serially but it's absolutely a point of concern.